sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala

```diff
@@ -0,0 +1,131 @@
+        GetDataFromCurrentView[SidechainHistory, SidechainState, SidechainWallet, SidechainMemoryPool, ForgeResult](forgingFunctionForEpochAndSlot)
+
+      forgeMessage
+  }
+
+  protected def tryToForgeNextBlock(nextConsensusEpochNumber: ConsensusEpochNumber, nextConsensusSlotNumber: ConsensusSlotNumber)(nodeView: View): ForgeResult = {
+    log.info(s"Try to forge block for epoch ${nextConsensusEpochNumber} with slot ${nextConsensusSlotNumber}")
+    val bestBlockId = nodeView.history.bestBlockId
+    val consensusInfo: FullConsensusEpochInfo = nodeView.history.getFullConsensusEpochInfoForNextBlock(bestBlockId, nextConsensusEpochNumber)
+    val sidechainWallet =  nodeView.vault
+
+    checkEpochAndSlotForgability(nodeView.history.bestBlockInfo.timestamp, nextConsensusEpochNumber, nextConsensusSlotNumber)
+
+    val forgerBoxMerklePathInfo: Seq[(ForgerBox, MerklePath)]
+      = sidechainWallet.getForgerBoxMerklePathInfoOpt(nextConsensusEpochNumber).getOrElse(Seq()).map(d => (d.forgerBox, d.merklePath))
+
+    val vrfMessage = buildVrfMessage(nextConsensusSlotNumber, consensusInfo.nonceConsensusEpochInfo)
+    val ownedForgingDataView: Seq[(ForgerBox, MerklePath, PrivateKey25519, VrfSecretKey, VrfProof)]
+      = forgerBoxMerklePathInfo.view.flatMap{case (forgerBox, merklePath) => getSecretsAndProof(sidechainWallet, vrfMessage, forgerBox, merklePath)}
+
+    val totalStake = consensusInfo.stakeConsensusEpochInfo.totalStake
+    val eligibleForgingDataView: Seq[(ForgerBox, MerklePath, PrivateKey25519, VrfSecretKey, VrfProof)] =
+      ownedForgingDataView.filter{case(forgerBox, merklePath, privateKey25519, vrfSecretKey, vrfProof) => vrfProofCheckAgainstStake(vrfProof, forgerBox.vrfPubKey(), vrfMessage, forgerBox.value(), totalStake)}
+
+    val eligibleForgerOpt = eligibleForgingDataView.headOption //force all forging related calculations
+
+    val nextBlockTimestamp = getTimeStampForEpochAndSlot(nextConsensusEpochNumber, nextConsensusSlotNumber)
+    val forgingResult = eligibleForgerOpt
+      .map{case (forgerBox, merklePath, privateKey25519, vrfSecretKey, vrfProof) =>
+        forgeBlock(nodeView, bestBlockId, nextBlockTimestamp, forgerBox, merklePath, privateKey25519, vrfSecretKey, vrfProof)}
+      .getOrElse(SkipSlot)
+
+    log.info(s"Forge result is: ${forgingResult}")
+    forgingResult
+  }
+
+  private def getSecretsAndProof(wallet: SidechainWallet, vrfMessage: VrfMessage, forgerBox: ForgerBox, merklePath: MerklePath) = {
+    for {
+      rewardPrivateKey <- wallet.secret(forgerBox.rewardProposition()).asInstanceOf[Option[PrivateKey25519]]
+      vrfSecret <- wallet.secret(forgerBox.vrfPubKey()).asInstanceOf[Option[VrfSecretKey]]
+    } yield (forgerBox, merklePath, rewardPrivateKey, vrfSecret, vrfSecret.prove(vrfMessage))
+  }
+
+
+  private def checkEpochAndSlotForgability(bestBlockTimestamp: Long, checkedEpochNumber: ConsensusEpochNumber, checkedSlotNumber: ConsensusSlotNumber): Unit = {
```

Comment:
Check method return type.
Actually do nothing, because resulting `ForgeFailed` is not processed by caller.